AP_Scripting: add set angle for copter guided

IamPete1 opened the pull request:
This adds access to guided mode set angle for scripting

rmackay9 commented:
This is great to see, I had been thinking of adding this myself.  It's a bit unusual to use a Vector3f for a roll, pitch, yaw angle though.  I think we should stick with three floats for the angles in degrees.

Really great though!

IamPete1 commented:
@rmackay9 fixed, I also set the scripting range check to +- 45 deg for roll and pitch and +-360 for yaw, is the too much of a constraint? I think there would be a 45deg hard limit anyway.

rmackay9 commented:
@IamPete1,

Great, thanks.  No need to pass them in as "const float&", just use floats I think.  A reference here takes up 4 bytes which is the same size as the float itself.

The yaw limit makes sense but I'd actually remove limits on roll and pitch.  As you say it should be limited internally by the ANGLE_MAX parameter but users are allowed to adjust that themselves.  I think in general we want users to have few restraints when doing scripting.

rmackay9 commented:
By the way, no pressure of course, but a team from the Japan drone school wants to try to implement drifting with a Rover using lua scripts.  They need a binding to give them direct control of steering and throttle..

IamPete1 commented:
@rmackay9 I have changed the limits to +-90 for roll and pitch and removed the references. I didn't know about the cost, I just duplicated what was done for the set velocity method. I guess in that case it is worth it because the vector3f is larger than the reference so its a net saving?

Overriding outputs from scripting went in recently, or were you thinking of a method that goes via guided mode?

rmackay9 commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -158,6 +158,7 @@ singleton AP_Vehicle method start_takeoff boolean float (-LOCATION_ALT_MAX_M*100
 singleton AP_Vehicle method set_target_location boolean Location
 singleton AP_Vehicle method get_target_location boolean Location'Null
 singleton AP_Vehicle method set_target_velocity_NED boolean Vector3f
+singleton AP_Vehicle method set_target_angle boolean float -90 90 float -90 90 float -360 360 float -FLT_MAX FLT_MAX boolean float -FLT_MAX FLT_MAX
```
The limit for roll should be -180 to +180 I think..

rmackay9 commented on ArduCopter/Copter.cpp:
```diff
@@ -309,6 +309,21 @@ bool Copter::set_target_velocity_NED(const Vector3f& vel_ned)
     return true;
 }
 
+bool Copter::set_target_angle(const float roll_deg, const float pitch_deg, const float yaw_deg, const float climb_rate_cms, const bool use_yaw_rate, const float yaw_rate_rads)
```
As a small side note, I'm really not a fan of adding "const" in front of arguments.  I know there are varying opinions on this but I'm personally not a fan because it makes the line much longer and harder to read 'cuz more words to ignore before getting to what matters.  This isn't a blocker though.

IamPete1 replied:
fixed

rmackay9 commented on libraries/AP_Vehicle/AP_Vehicle.h:
```diff
@@ -170,10 +170,12 @@ class AP_Vehicle : public AP_HAL::HAL::Callbacks {
     virtual bool start_takeoff(float alt) { return false; }
     virtual bool set_target_location(const Location& target_loc) { return false; }
     virtual bool set_target_velocity_NED(const Vector3f& vel_ned) { return false; }
+    virtual bool set_target_angle(const float roll_deg, const float pitch_deg, const float yaw_deg, const float climb_rate_cms, const bool use_yaw_rate, const float yaw_rate_rads) { return false; }
```
instead of "set_target_angle" it might be "set_target_angle_and_climbrate" but again not a blocker.

IamPete1 replied:
fixed

rmackay9 commented:
@WickedShell are you happy with this?  I am besides a couple of minor points.  I think it's an important feature to have for Copter and it's inline with what companion computers can do already with the set-target-attitude message.

rmackay9 commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -164,6 +164,7 @@ singleton AP_Vehicle method start_takeoff boolean float (-LOCATION_ALT_MAX_M*100
 singleton AP_Vehicle method set_target_location boolean Location
 singleton AP_Vehicle method get_target_location boolean Location'Null
 singleton AP_Vehicle method set_target_velocity_NED boolean Vector3f
+singleton AP_Vehicle method set_target_angle_and_climbrate boolean float -90 90 float -180 180 float -360 360 float -FLT_MAX FLT_MAX boolean float -FLT_MAX FLT_MAX
```
so close!  I think the roll limits should be +-180 and pitch should be +-90.  I think it's reversed here.

IamPete1 replied:
oops, fixed, hopefully I will get there in the end

rmackay9 commented on ArduCopter/Copter.cpp:
```diff
@@ -309,6 +309,21 @@ bool Copter::set_target_velocity_NED(const Vector3f& vel_ned)
     return true;
 }
 
+bool Copter::set_target_angle_and_climbrate(float roll_deg, float pitch_deg, float yaw_deg, float climb_rate_cms, bool use_yaw_rate, float yaw_rate_rads)
```
sorry to keep drip feeding changes but I suspect that we should make the climb rate in m/s instead of cm/s.

IamPete1 replied:
yaw rate in deg per second also? to match roll, pitch and yaw

rmackay9 replied:
Yes, that makes sense too.  good catch.

IamPete1 replied:
fixed units

rmackay9 commented on ArduCopter/Copter.h:
```diff
@@ -635,6 +635,7 @@ class Copter : public AP_Vehicle {
     bool start_takeoff(float alt) override;
     bool set_target_location(const Location& target_loc) override;
     bool set_target_velocity_NED(const Vector3f& vel_ned) override;
+    bool set_target_angle_and_climbrate(float roll_deg, float pitch_deg, float yaw_deg, float climb_rate_cms, bool use_yaw_rate, float yaw_rate_rads) override;
```
looks like it's got the argument named as "climb_rate_cms"

IamPete1 replied:
I'm surprised this build at all

rmackay9 replied:
Yeah, C++ only checks that the types match, not that the names of the arguments match.

rmackay9 commented:
I'm keen that we get this in before the weekend if possible because I have some drone school students here in Japan who would like to make use of this.

I think it's quite a straightforward change.
